content/blog/2024-10-21-testing-the-limits/index.md

```diff
@@ -0,0 +1,242 @@
+  valkey-node-3:
+    hostname: valkey3
+    image: valkey/valkey:latest
+    command: valkey-server --port 6381 --cluster-enabled yes --cluster-config-file nodes.conf --cluster-node-timeout 5000 --requirepass ${VALKEY_PASSWORD}
+    volumes:
+      - ./data3:/data
+    network_mode: host
+
+volumes:
+  data1:
+    driver: local
+  data3:
+    driver: local
+  data2:
+    driver: local
+```
+Run `docker compose -f valkey-cluster.yaml up -d` to start the cluster. There is one more step to get the cluster running. Find the name of one of your nodes with `docker ps --format '{{.Names}}'`. 
+
+```bash
+docker ps --format '{{.Names}}
+kvtest-valkey-node-1-1
+kvtest-valkey-node-3-1
+kvtest-valkey-node-2-1
+```
+
+I'll use the first container to finish the cluster creation. Once the containers have started up we have to tell them the details they need to use for the cluster. Below I am using the IP of the host and the port configurations of all the containers to create the cluster. This is because these addresses need to be accessible from the benchmarking server. 
+
+```bash
+docker exec -it kvtest-valkey-node-1-1 valkey-cli --cluster create 10.0.1.136:6379 10.0.1.136:6380 10.0.1.136:6381 -a e41fb9818502071d592b36b99f63003019861dad
+```
+
+Now we can run our benchmark! We need to add the `--cluster` flag to our benchmarking command. Also, because this is so fast I ended up moving from 1 million requests to 10 million requests. That way we can make sure Valkey has time to fully utilize all it's resources. 
+
+```bash
+redis-benchmark -n 10000000 -t set,get -P 16 -q --threads 10 --cluster -a e41fb9818502071d592b36b99f63003019861dad --threads 5 -h 10.0.1.136 
+Cluster has 3 master nodes:
+
+Master 0: 219294612b44226fa32482871cf21025ff531875 10.0.1.136:6380
+Master 1: e5d85b970551c27065f1552f5358f4add6114d98 10.0.1.136:6381
+Master 2: 1faf3d0dd22e518eec11fd46c0de6ce18cd15cfe 10.0.1.136:6379
+
+SET: 1122838.50 requests per second, p50=0.575 msec                     
+GET: 1188071.75 requests per second, p50=0.511 msec  
+```
+
+**1,155,000 requests per second**. We've managed to double our requests per second. All this on a single board computer that's the size of a credit card. 
+
+While this is far from what I would recommend for a production server these are the same steps I'd reccomend to someone evaluating Valkey. It's important test with a single instance to start finding the optimal settings. Then you can begin to scale up your test by adding either more IO Threads or Valkey instances. 
+
+Testing should mirror your production workload as best it can. This test is using syntehtic data. That's why I'd recommend checking out the documentation to find what other settings you may need to test with. For example, we tested with the default settings of 50 client connections and 3 byte payloads. Your production workload may look different so explore all the settings! You may find that IO threading works better for your use case than I did in this case.
```

Comment:
```suggestion
Testing should mirror your production workload as best it can. This test is using synthetic data. That's why I'd recommend checking out the documentation to find what other settings you may need to test with. For example, we tested with the default settings of 50 client connections and 3 byte payloads. Your production workload may look different so explore all the settings! You may find that IO threading works better for your use case than I did in this case.
```